File: November 2022/03 November 2022/Python/Array Removals.py
#User function Template for python3


class Solution:

	def removals(self,arr, n, k):
        arr.sort()
        ans = float('inf')
        
        for i in range(len(arr)):
            lo, hi = 0, i+1
            while lo < hi:
                mi = lo+(hi-lo)//2
                if arr[i]-arr[mi] <= k:
                    hi = mi
                else:
                    lo = mi+1
            
            if lo > i:
                continue
            ans = min(ans, n-(i-lo+1))
           # print("i = ", i, "lo = ", lo, " ans = ", ans)
        return ans


#{ 
 # Driver Code Starts
#Initial Template for Python 3


if __name__ == '__main__':
    tc = int(input())
    while tc > 0:
        n, k = list(map(int, input().strip().split()))
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.removals(arr, n, k)
        print(ans)
        tc -= 1
# } Driver Code Ends

File: November 2022/03 November 2022/C++/Array Removals.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution{
    public:
    int removals(vector<int>& arr, int k){
        //Code here
        int n = arr.size();
        sort(arr.begin(),arr.end());  //O(NLOGN)
        int ans = n-1; //at max there can be n-1 removals
        for(int i=0; i<n; i++){
            int j = upper_bound(arr.begin()+i,arr.begin()+n,arr[i]+k)-arr.begin()-1;
            ans = min(ans,n-(j-i+1));
        }
        return ans;
    }
};


//{ Driver Code Starts.


int main(){
    int t;
    cin>>t;
    
    while(t--){
        int n,k;
        cin>>n>>k;
        vector<int> arr(n);
        for(int i=0;i<n;i++){
            cin>>arr[i];
        }
        
        Solution ob;
        int ans = ob.removals(arr,k);
        
        cout<<ans<<endl;
    }
}



// } Driver Code Ends